ruby/ql/lib/codeql/ruby/security/ZipSlipCustomizations.qll

```diff
@@ -0,0 +1,125 @@
+/**
+ * Provides default sources, sinks and sanitizers for reasoning about
+ * zip slip vulnerabilities, as well as extension points for
+ * adding your own.
+ */
+
+private import codeql.ruby.AST
+private import codeql.ruby.ApiGraphs
+private import codeql.ruby.CFG
+private import codeql.ruby.Concepts
+private import codeql.ruby.DataFlow
+private import codeql.ruby.dataflow.BarrierGuards
+private import codeql.ruby.dataflow.RemoteFlowSources
+
+/**
+ * Provides default sources, sinks and sanitizers for reasoning about
+ * zip slip vulnerabilities, as well as extension points for
+ * adding your own.
+ */
+module ZipSlip {
+  /**
+   * A data flow source for zip slip vulnerabilities.
+   */
+  abstract class Source extends DataFlow::Node { }
+
+  /**
+   * A data flow sink for zip slip vulnerabilities.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A sanitizer for zip slip vulnerabilities.
+   */
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * A file system access, considered as a flow sink.
+   */
+  class FileSystemAccessAsSink extends Sink {
+    FileSystemAccessAsSink() { this = any(FileSystemAccess e).getAPathArgument() }
+  }
+
+  /**
+   * A call to `Zlib::GzipReader.open(path)`, considered a flow source.
+   */
+  private class GzipReaderOpen extends Source {
+    GzipReaderOpen() {
+      this = API::getTopLevelMember("Zlib").getMember("GzipReader").getReturn("open").asSource() and
+      // If argument refers to a string object, then it's a hardcoded path and
+      // this file is safe.
+      not this.(DataFlow::CallNode)
+          .getArgument(0)
+          .getALocalSource()
+          .getConstantValue()
+          .isStringlikeValue(_)
+    }
+  }
+
+  /**
+   * A call to `Gem::Package::TarReader.new(file_stream)`, considered a flow source.
+   */
+  private class TarReaderInstance extends Source {
+    TarReaderInstance() {
+      this =
+        API::getTopLevelMember("Gem")
+            .getMember("Package")
+            .getMember("TarReader")
+            .getInstance()
+            .asSource() and
+      // If argument refers to a string object, then it's a hardcoded path and
+      // this file is safe.
+      not this.(DataFlow::CallNode)
+          .getArgument(0)
+          .getALocalSource()
+          .getConstantValue()
+          .isStringlikeValue(_)
+    }
+  }
+
+  /**
+   * A call to `Zip::File.open(path)`, considered a flow source.
+   */
+  private class ZipFileOpen extends Source {
```

Comment:
The following project resulted in some strange looking results:
https://gist.github.com/aibaars/9f25acb1dcc81b4731d9781b955d3ddd

I think you need to model the case where the `Zip::File.open` method is called with a block differently from the case where it is called without block;  `Zip::File.open(path) do |file| ... end` and `file = Zip::File.open(path)` are different.